File: web/src/lib/leaderboardPeriod.ts
// Shared leaderboard period model — web + iOS keep the same 3-case
// canonical set: This Week / This Month / All Time, all rolling
// (now - 7d, now - 30d, null). Matches `LeaderboardPeriod.swift`.
//
// Inline `d.setDate(d.getDate() - 30)` math + iOS calendar-bucket
// (`yearForWeekOfYear`, `weekOfYear`) was inconsistent across surfaces;
// this helper is the single source of truth.

export const PERIOD_LABELS = ['This Week', 'This Month', 'All Time'] as const;
export type Period = (typeof PERIOD_LABELS)[number];

const DAY_MS = 86_400_000;
const WINDOW_DAYS: Record<Period, number | null> = {
  'This Week': 7,
  'This Month': 30,
  'All Time': null,
};

/**
 * Rolling cutoff for a leaderboard window. Returns `null` for All Time
 * (caller should skip the cutoff filter entirely).
 */
export function periodSince(period: Period, now: Date = new Date()): Date | null {
  const days = WINDOW_DAYS[period];
  if (days == null) return null;
  return new Date(now.getTime() - days * DAY_MS);
}

File: VerityPost/VerityPost/LeaderboardPeriod.swift
import Foundation

// Shared leaderboard period model — web + iOS keep the same 3-case
// canonical set: This Week / This Month / All Time, all rolling
// (now - 7d, now - 30d, nil). Mirrors `web/src/lib/leaderboardPeriod.ts`.
//
// Replaces the prior calendar-bucket math
// (`Calendar.dateComponents([.yearForWeekOfYear, .weekOfYear], …)`)
// which produced different "This Week" answers from the web's rolling
// 7-day window.

enum LeaderboardPeriod: String, CaseIterable, Identifiable {
    case thisWeek = "This week"
    case thisMonth = "This month"
    case allTime = "All time"

    var id: String { rawValue }

    /// Rolling cutoff for the window. Returns `nil` for `.allTime`
    /// (caller should skip the date filter entirely).
    func since(date now: Date = Date()) -> Date? {
        switch self {
        case .thisWeek:  return now.addingTimeInterval(-7 * 86_400)
        case .thisMonth: return now.addingTimeInterval(-30 * 86_400)
        case .allTime:   return nil
        }
    }
}
